Implement Ajtai commitment scheme with lattice-based cryptography

pycckuu opened the pull request:
closes https://github.com/NethermindEth/condor-rs/issues/6

## Summary
Added Ajtai commitment scheme implementation using lattice-based cryptography with parameter validation, matrix-vector operations, and cryptographic security checks. Introduces core cryptographic primitives for polynomial ring operations and commitment verification.

## Code Changes
1. **New Modules**:
   - `ajtai_commitment.rs`: Core commitment logic with security validation
   ```rust
   impl<const M: usize, const N: usize, const D: usize> AjtaiCommitment<M, N, D> {
       fn verify_security_relation(beta: u32, m: u128, q: u128) -> Result<(), ParameterError> {
           // Security condition: β²m³ < q²
       }
   }
   ```
   - `rq_matrix.rs`: matrix-vector multiplication
   ```rust
   impl<const M: usize, const N: usize, const D: usize> RqMatrix<M, N, D> {
       pub fn mul_vec(&self, vec: &[Rq<D>; N]) -> [Rq<D>; M] {
           // matrix multiplication
       }
   }
   ```

2. **Polynomial Operations**:
   - Added bounded random generation and message encoding in `rq.rs`
   ```rust
   impl<const D: usize> Rq<D> {
       pub fn random_small() -> Self {
           // Generates coefficients in {-1, 0, 1}
       }
   }
   ```


## Test Plan
**Verified**:
1. Parameter validation rejects insecure configurations:
   ```rust
   #[test]
   fn rejects_invalid_parameters() {
       assert!(AjtaiParameters::new(Zq::one(), Zq::zero(), Zq::one()).is_err());
   }
   ```
2. Commitment binding through 100-operation stress test
3. Correct verification of tampered openings

## Additional Notes
-  Uses thread-local RNG for randomness generation
-  Implements optimized matrix multiplication with chunking
-  Includes error handling with 6 distinct error types

v0-e commented:
I glanced over the PR and it pretty cool @pycckuu, pretty nice overall :100:
Just a couple of tips,
- it would be nice to have the ability to choose which RNG is used, be it for `no-std` reasons, be able to choose the seed, choose RNGs with more security, etc., some function like `fn(rng: &mut impl Rng) -> `;
- we will most likely be dealing with proof systems with a large number of constraints (~ 2^20), so don't be afraid to use vectors 😄. Stuff like a `[[T; N]; M]` will most likely blow up the stack, considering if at least one of its dimensions will be around the number of constraints.

mattsuffern commented:
Awesome work!! I'm still going through the details and trying to fully understand everything. This is more of a question than a comment:

- If $\lVert s_{i} \rVert \leq \beta$, what is the difference between `beta` and `witness_bound`?

pycckuu commented:
>  what is the difference between `beta` and `witness_bound`? (AjtaiParameters)

@mattsuffern  As I understand it:
-	⁠beta:  Validates the theoretical security of the scheme design. It's a higher-level parameter used in security proofs and parameter validation.
-	⁠witness_bound and ⁠randomness_bound: Enforce practical constraints on the implementation.

maksimryndin commented on labrador/src/rq.rs:
```diff
@@ -118,6 +119,68 @@ impl<const D: usize> Rq<D> {
+
+    /// Generate random small polynomial for commitments
+    pub fn random_small() -> Self {
+        let mut rng = rand::rng();
```
It would be nice to have a type-enforced requirement for rng to implement https://docs.rs/rand/latest/rand/trait.CryptoRng.html  trait in order to avoid any sudden breach because of the changes in the `rand` crate.

E.g.

```rust
let rng: &mut dyn rand::CryptoRng = &mut rand::rng();
```

pycckuu replied:
I am not sure I got your point on how it can be breached. Here is my proposed fix fed1eac

maksimryndin replied:
Yes, it works! The point is that we should have a `CryptoRng` set explicitly as a trait bound. 
The point was that if in some future versions of `rand` crate `rand::rng()` may start returning a non-cryptographic rng and a blind update of dependencies would lead to a security breach.

maksimryndin reviewed (commented):
Awesome, @pycckuu ! 👍 

I really like a high quality of your work 🥇 
Some small comments and nit pick, will review again after

NiDimi commented on labrador/src/ajtai_commitment.rs:
```diff
@@ -0,0 +1,287 @@
+    pub fn new(params: AjtaiParameters) -> Result<Self, ParameterError> {
+        Self::validate_parameters(&params)?;
+        Ok(Self {
+            matrix_a: RqMatrix::random(),
```
Yes the random matrix here represents the public key A in the commitment.

maksimryndin replied:
Should `matrix_a` come outside `AjtaiCommitment`? At least for the purposes of a transcript?

NiDimi replied:
It might make sense to have two methods. 1) Try intialiasing the AjtaiCommitment with a matrix as an argument 2) Get a random AjtaiCommitment

NiDimi commented on labrador/src/ajtai_commitment.rs:
```diff
@@ -0,0 +1,287 @@
+        let randomness = self.generate_bounded_randomness()?;
+
+        let mut commitment = self.matrix_a.mul_vec(&witness);
+        Self::add_randomness(&mut commitment, &randomness);
```
@pycckuu I am a bit confused on why we need to add randomness for the commitment. As far as I am aware we should have:
$t = A \cdot s$ 
where `A` is the public matrix, and `s` is the message vector

mattsuffern reviewed (commented):
Awesome work! Everything looks good to me, and I don’t have any changes to suggest

pycckuu commented on labrador/src/ajtai_commitment.rs:
```diff
@@ -0,0 +1,287 @@
+impl AjtaiParameters {
+    /// Creates new parameters with validation
+    pub fn new(beta: Zq, witness_bound: Zq, randomness_bound: Zq) -> Result<Self, ParameterError> {
+        if witness_bound.is_zero() || randomness_bound.is_zero() {
```
After our discussion, we are removing "hiding" parameter `r`, therefore we're now only constrained by the witness bound.

NiDimi commented:
@pycckuu Can you share the resources you used for your implementation?

pycckuu commented:
> @pycckuu Can you share the resources you used for your implementation?

Nothing specific. On top of what was previously shared in Slack, I've researched existing implementations including Lazarus and the ajtai crate. I also utilized various resources such as papers, online materials, and YouTube videos. Some specific references I consulted (among others):

1. Ajtai 1996
2. Concretely Efficient Lattice-based Polynomial Commitment from Standard Assumptions 2024
3. https://crates.io/crates/ajtai
4. https://www.math.uci.edu/~asilverb/CryptoDayFiles/uci15a.pdf
5. https://www.youtube.com/watch?v=eHMyplex7YQ

pycckuu commented:
> I glanced over the PR and it looks pretty cool @pycckuu, pretty nice overall 💯 Just a couple of tips,
> 
> * it would be nice to have the ability to choose which RNG is used, be it for `no-std` reasons, be able to choose the seed, choose RNGs with more security, etc., some function like `fn(rng: &mut impl Rng) -> `;
> * we will most likely be dealing with proof systems with a large number of constraints (~ 2^20), so don't be afraid to use vectors 😄. Stuff like a `[[T; N]; M]` matrix will most likely blow up the stack, considering if at least one of its dimensions will be around the number of constraints.

addressed in 02df78c and 54486b0

pycckuu commented:
@v0-e @mattsuffern @NiDimi @maksimryndin I think I addressed all the comments. Please take a look again.

maksimryndin reviewed (approved):
LGTM! 👍

pycckuu commented:
> @v0-e @mattsuffern @NiDimi @maksimryndin I think I addressed all the comments. Please take a look again.

Could you kindly review and approve the changes for the PR merge when you have a moment?

v0-e commented on labrador/src/ajtai_commitment.rs:
```diff
@@ -0,0 +1,292 @@
+    ZeroParameter,
+    #[error("security bound β·m^(3/2) must be less than q")]
+    SecurityBoundViolation,
+    #[error("invalid witness bounds specified")]
+    InvalidWitnessBounds,
+}
+
+#[derive(Debug, Error)]
+pub enum CommitError {
+    #[error("witness coefficients exceed bound")]
+    WitnessBoundViolation,
+}
+
+#[derive(Debug, Error)]
+pub enum VerificationError {
+    #[error("witness coefficients exceed bound")]
+    WitnessBoundViolation,
```
For some errors it is cool to have the value that led to the error, eg.,
```rust
#[error("value {0} is too large")]
OutOfBounds(u64),
```

maksimryndin replied:
addressed https://github.com/NethermindEth/condor-rs/pull/21/commits/25af8bfa2e85045d72adee4695089875a2efba6a

v0-e commented on labrador/src/ajtai_commitment.rs:
```diff
@@ -0,0 +1,292 @@
+    }
+}
+
+impl Default for AjtaiParameters {
+    fn default() -> Self {
+        Self {
+            beta: Zq::one(),
+            witness_bound: Zq::one(),
+        }
+    }
+}
```
Didn't check, is `1` the correct default value? If default does not exist for this we can remove the `Default` impl.

maksimryndin replied:
addressed [9e49b8a94817f1a2f9b4e14165ab82678a3e55b1](https://github.com/NethermindEth/condor-rs/pull/21/commits/9e49b8a94817f1a2f9b4e14165ab82678a3e55b1)

v0-e commented on labrador/src/ajtai_commitment.rs:
```diff
@@ -0,0 +1,292 @@
+}
+
+impl AjtaiParameters {
+    /// Creates new parameters with validation
+    pub fn new(beta: Zq, witness_bound: Zq) -> Result<Self, ParameterError> {
+        if witness_bound.is_zero() {
+            return Err(ParameterError::InvalidWitnessBounds);
+        }
+
+        Ok(Self {
+            beta,
+            witness_bound,
+        })
+    }
```
Personally I like to avoid fallible `new` constructors, though can be a thing for other structs.

maksimryndin replied:
Let's leave it for now as-is and then see the use cases :)

v0-e commented on labrador/src/ajtai_commitment.rs:
```diff
@@ -0,0 +1,292 @@
+    }
+}
+
+/// Cryptographic opening containing witness
+#[derive(Clone, Debug)]
+pub struct Opening<const N: usize, const D: usize> {
+    pub witness: [Rq<D>; N],
+}
```
Maybe this should be a vector / use the wrapper, not sure!

maksimryndin replied:
addressed https://github.com/NethermindEth/condor-rs/pull/21/commits/25af8bfa2e85045d72adee4695089875a2efba6a

v0-e commented on labrador/src/rq.rs:
```diff
@@ -118,6 +119,96 @@ impl<const D: usize> Rq<D> {
+        Self::random_small_with_any_rng(&mut rng)
+    }
+
+    /// Generate random small polynomial with a provided cryptographically secure RNG
+    pub fn random_small_with_secure_rng<R: Rng + CryptoRng>(rng: &mut R) -> Self {
+        Self::random_small_with_any_rng(rng)
+    }
```
I would have this signature as the "default" `random_small()` function.

v0-e commented on labrador/src/rq_matrix.rs:
```diff
@@ -0,0 +1,63 @@
+
+    /// Matrix-vector multiplication that returns an array
+    pub fn mul_vec(&self, vec: &[Rq<D>; N]) -> [Rq<D>; M] {
+        (self * &RqVector::from(vec.clone())).into_array()
```
Can we avoid the cloning here?

maksimryndin replied:
addressed https://github.com/NethermindEth/condor-rs/pull/21/commits/25af8bfa2e85045d72adee4695089875a2efba6a

v0-e commented on labrador/src/rq_vector.rs:
```diff
@@ -0,0 +1,95 @@
+    }
+}
+
+// Dot product between two vectors
+impl<const N: usize, const D: usize> Mul for &RqVector<N, D> {
+    type Output = Rq<D>;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        self.elements
+            .iter()
+            .zip(rhs.elements.iter())
+            .map(|(a, b)| a.clone() * b.clone())
+            .fold(Rq::zero(), |acc, x| acc + x)
+    }
+}
```
If we have ownership of `self`, and `rhs` we can avoid the cloning here.
`MulAssign` may also be helpful.

maksimryndin replied:
addressed https://github.com/NethermindEth/condor-rs/pull/21/commits/25af8bfa2e85045d72adee4695089875a2efba6a

v0-e reviewed (changes requested):
Looks good overall! :)

mattsuffern reviewed (approved):
Everything looks good! No issues found. Great work!

maksimryndin commented on labrador/src/ajtai_commitment.rs:
```diff
@@ -0,0 +1,292 @@
+    /// - q is the modulus of the underlying ring
+    fn verify_security_relation(beta: u32, m: u128) -> Result<(), ParameterError> {
+        // Calculate q from Zq properties
+        let q_val = (Zq::zero() - Zq::one()).value();
```
addressed https://github.com/NethermindEth/condor-rs/pull/21/commits/25af8bfa2e85045d72adee4695089875a2efba6a

maksimryndin commented on labrador/src/rq.rs:
```diff
@@ -118,6 +119,96 @@ impl<const D: usize> Rq<D> {
+    }
+
+    /// Check if polynomial coefficients are within bounds
+    pub fn check_bounds(&self, bound: Zq) -> bool {
```
Could you please elaborate why having zero bound is a problem here for `Rq`?

NiDimi replied:
If bound = 0 then we have the condition 
`val <= bound.value() || val >= Zq::Q.wrapping_sub(bound.value())`
if val <=0 is true only when val is 0 because it cannot be negative and `Zq::Q.wrapping_sub(0) = 0` so we end up having the condition `val <= 0 || val >= 0`. I don't think its a big problem just thought I should make you aware.

maksimryndin replied:
So, it is `[-bound, bound] := [0, 0]`, right? Could it be a problem somewhere?
For now I cannot catch up why it is an edge case here (as we consider closed interval and we have zero-coefficient polynomials which fit it. And the method returns boolean)

The improvement I see here is to implement comparison operations at `Zq` in order not to leak the abstraction. created the issue https://github.com/NethermindEth/condor-rs/issues/32

NiDimi replied:
I dont see yet where this could be a problem but is nice to document stuff like this in case it becomes a problem :sweat_smile:

maksimryndin replied:
Okay :) thanks for spotting that :)

v0-e reviewed (approved):
Nice! :)
